src/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,163 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+#include "src/common/utils.hpp"
+#include <functional>
+#include <iostream>
+#include <measurement_kit/common.hpp>
+#include <measurement_kit/ext.hpp>
+#include <measurement_kit/http.hpp>
+#include <measurement_kit/neubot.hpp>
+#include <stdlib.h>
+#include <string>
+#include <unistd.h>
+
+#define DASH_MAX_ITERATION 15
+#define DASH_SECONDS 2
+
+using namespace mk;
+using namespace mk::net;
+using namespace mk::http;
+using json = nlohmann::json;
+
+namespace mk {
+namespace neubot {
+namespace dash {
+
+static inline void loop_request(Var<Transport> transport, int speed_kbit,
+                                Callback<Error, Var<json>> cb, std::string auth,
+                                Var<json> measurements, Var<Reactor> reactor,
+                                Var<Logger> logger, int iteration = 1) {
+
+    int rate_index = 0;
+    Settings settings;
+    int DASH_RATES[] = {100,  150,  200,  250,  300,   400,  500,
+                        700,  900,  1200, 1500, 2000,  2500, 3000,
+                        4000, 5000, 6000, 7000, 10000, 20000};
+    std::string url = "http://127.0.0.1/dash/download/";
+
+    if (iteration > DASH_MAX_ITERATION) {
+        transport->close([=]() { cb(NoError(), measurements); });
+        return;
+    };
+
+    // Bisect
+    while (DASH_RATES[rate_index] < speed_kbit) {
+        rate_index++;
```

Comment:
I think here you should check not to go beyond the vector's boundaries, shouldn't you?